timelock-impl/src/test/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionDurationAccumulatorTest.java

```diff
@@ -0,0 +1,152 @@
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 5, 1);
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_3, 5, 7);
+
+        verify(mockConsumer).accept(10L);
+        verify(mockConsumer).accept(15L);
+        verify(mockConsumer).accept(7L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    @SuppressWarnings("ExecutorSubmitRunnableFutureIgnored")
+    public void testInterleavingNoPause() throws InterruptedException {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(3);
+        accumulator = new LeaderElectionDurationAccumulator(mockConsumer, 300);
+        executorService.submit(() -> leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 300, 1));
+        executorService.submit(() -> leaderElectionResultsWithDurationInRandomOrder(LEADER_2, LEADER_3, 300, 17));
+        executorService.submit(() -> leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_3, 300, 76));
+        executorService.shutdown();
+        executorService.awaitTermination(1, TimeUnit.MINUTES);
+
+        verify(mockConsumer).accept(1L);
+        verify(mockConsumer).accept(17L);
+        verify(mockConsumer).accept(76L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    @SuppressWarnings("ExecutorSubmitRunnableFutureIgnored")
+    public void testInterleavingWithPause() throws InterruptedException {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(3);
+        accumulator = new LeaderElectionDurationAccumulator(mockConsumer, 300);
+        executorService.submit(() -> leaderElectionResultsWithPause(LEADER_1, LEADER_2, 300, 1));
+        executorService.submit(() -> leaderElectionResultsWithPause(LEADER_2, LEADER_3, 300, 17));
+        executorService.submit(() -> leaderElectionResultsWithPause(LEADER_1, LEADER_3, 300, 76));
+        executorService.shutdown();
+        executorService.awaitTermination(1, TimeUnit.MINUTES);
+
+        verify(mockConsumer).accept(1L);
+        verify(mockConsumer).accept(17L);
+        verify(mockConsumer).accept(76L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    @SuppressWarnings("ExecutorSubmitRunnableFutureIgnored")
+    public void testManyUpdatesForSameLeaders() throws InterruptedException {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(50);
+        accumulator = new LeaderElectionDurationAccumulator(mockConsumer, 5_000);
+        List<Integer> durationBuckets =
+                IntStream.range(0, 100).map(x -> x * 100).boxed().collect(Collectors.toList());
+        Collections.shuffle(durationBuckets);
+
+        durationBuckets.forEach(duration ->
+                executorService.submit(() -> leaderElectionResultsWithPause(LEADER_1, LEADER_2, 50, duration)));
+
+        executorService.shutdown();
+        executorService.awaitTermination(1, TimeUnit.MINUTES);
+
+        verify(mockConsumer).accept(0L);
```

Comment:
~I'm possibly confused: how do we guarantee zero is processed in the first five~

Ah, this test has an accumulator with 5000. What you have is fine, don't really like that it relies on 5000 being 50 * 100 (maybe introduce some local variables so that it's obvious?) / if you change line 129 to 60 the test breaks, so maybe we could change that slightly

Reply:
Factored out two variables here, should make it clearer.